src/services/factories/getTrades.ts

```diff
@@ -2,42 +2,114 @@ import Web3 from 'web3'
 
-import ExchangeApiImpl, { ExchangeApi, Trade, Order, AuctionElement } from 'api/exchange/ExchangeApi'
+import ExchangeApiImpl, {
+  ExchangeApi,
+  Trade,
+  TradeReversion,
+  Order,
+  AuctionElement,
+  BaseTradeEvent,
+} from 'api/exchange/ExchangeApi'
 
 import { getTokensFactory } from 'services/factories/tokenList'
 import { addUnlistedTokensToUserTokenListByIdFactory } from 'services/factories/addUnlistedTokensToUserTokenListById'
 
 import { dateToBatchId, calculateSettlingTimestamp, logDebug, isOrderDeleted } from 'utils'
 
-interface GetTradesParams {
+interface GetTradesAndTradeReversionParams {
   networkId: number
   userAddress: string
   fromBlock?: number
   toBlock?: number | 'latest' | 'pending'
   orders: AuctionElement[]
 }
 
-export function getTradesFactory(factoryParams: {
+interface GetTradesAndTradeReversionReturn {
+  trades: Trade[]
+  reverts: TradeReversion[]
+}
+
+export function getTradesAndTradeReversionsFactory(factoryParams: {
   web3: Web3
   exchangeApi: ExchangeApi
   getTokens: ReturnType<typeof getTokensFactory>
   addUnlistedTokensToUserTokenListById: ReturnType<typeof addUnlistedTokensToUserTokenListByIdFactory>
-}): (params: GetTradesParams) => Promise<Trade[]> {
+}): (params: GetTradesAndTradeReversionParams) => Promise<GetTradesAndTradeReversionReturn> {
   const { web3, exchangeApi, getTokens, addUnlistedTokensToUserTokenListById } = factoryParams
 
   async function getBlockTimePair(blockNumber: number): Promise<[number, number]> {
     // js timestamp == unix timestamp * 1000
     return [blockNumber, +(await web3.eth.getBlock(blockNumber)).timestamp * 1000]
   }
 
-  async function getTrades(params: GetTradesParams): Promise<Trade[]> {
+  async function assembleTradeReversions(
+    events: BaseTradeEvent[],
+    blockTimes: Map<number, number>,
+  ): Promise<TradeReversion[]> {
+    return events.map<TradeReversion>(event => {
+      const timestamp = blockTimes.get(event.blockNumber) as number
+      const batchId = dateToBatchId(timestamp)
+      const revertKey = ExchangeApiImpl.buildTradeRevertKey(batchId, event.orderId)
+
+      return { ...event, timestamp, batchId, revertKey }
+    })
+  }
+
+  async function assembleTrades(
+    events: BaseTradeEvent[],
+    blockTimes: Map<number, number>,
+    orders: Map<string, Order>,
+    tokens: Map<number, TokenDetails>,
+  ): Promise<Trade[]> {
+    return events.map<Trade>(event => {
+      const timestamp = blockTimes.get(event.blockNumber) as number
+      const batchId = dateToBatchId(timestamp)
+      const revertKey = ExchangeApiImpl.buildTradeRevertKey(batchId, event.orderId)
+
+      const buyToken = tokens.get(event.buyTokenId) as TokenDetails
+      const sellToken = tokens.get(event.sellTokenId) as TokenDetails
+
+      // Maybe we couldn't find the order /shrug
+      const order = orders.get(event.orderId)
+
+      return {
+        ...event,
+        batchId,
+        timestamp,
+        revertKey,
```

Comment:
I don't think this should be part of the data object, right?
this seems more like an internal thing we use for calculating things, so we don't want to expose it.

Reply:
Why not? And which one specifically are you talking about?

Reply:
I mean `revertKey` looks like a derived field, and implementation detail not part of the trade data.

Imo, It doesn't belong in any component, service, or hook. Maybe is just part of the reducer, and maybe not even that is required if we consider other ways to update the trades. 

Also, not a big deal, and nitpick so if it makes it too complicated, ignore it

Reply:
Ah yes, this makes sense. Removed.